Python/import.c

```diff
@@ -308,6 +309,89 @@ _PyImport_IsInitialized(PyInterpreterState *interp)
     return 1;
 }
 
+PyObject *
+_PyImport_GetModule(PyObject *name)
+{
+    PyObject *modules = PyImport_GetModuleDict();
+    if (PyDict_CheckExact(modules)) {
+        return PyDict_GetItem(modules, name);
+    }
+
+    PyObject *mod = PyObject_GetItem(modules, name);
+    // For backward-comaptibility we copy the behavior of PyDict_GetItem().
+    if (PyErr_Occurred()) {
+        PyErr_Clear();
+    }
+    Py_XDECREF(mod);
+    return mod;
+}
+
+PyObject *
+_PyImport_GetModuleWithError(PyObject *name)
+{
+    PyObject *modules = PyImport_GetModuleDict();
+    if (PyDict_CheckExact(modules)) {
+        return PyDict_GetItemWithError(modules, name);
+    }
+
+    PyObject *mod = PyObject_GetItem(modules, name);
+    // For backward-comaptibility we copy the behavior
+    // of PyDict_GetItemWithError().
+    if (PyErr_ExceptionMatches(PyExc_KeyError)) {
+        PyErr_Clear();
+    }
+    return mod;
```

Comment:
This is missing the Py_XDECREF to return a borrowed reference instead of a new one.

Reply:
Gah! The docs for PyDict_GetItemWithError() before 3.7 don't mention it returning a borrowed reference. Thanks for catching that. I was looking for that refleak!

Reply:
Hmm.  Looks like that wasn't the refleak I was looking for. :/  Back I go.

Reply:
I was just pattern matching with the code above based on the assumption that `PyDict_GetItem` and `PyDict_GetItemWithError` exposed the same refcount semantics :)

However, it now occurs to me that these access patterns are thoroughly dubious if you genuinely want to allow dict subclasses and arbitrary mappings, as those may return *dynamic* references (e.g. from `__missing__` methods) , in which case the `Py_XDECREF()` line may invalidate the reference entirely.

Reply:
Good call.  I'll fix that.